src/hotspot/share/oops/compressedKlass.cpp

```diff
@@ -77,60 +192,105 @@ char* CompressedKlassPointers::reserve_address_space_X(uintptr_t from, uintptr_t
 }
 
 char* CompressedKlassPointers::reserve_address_space_for_unscaled_encoding(size_t size, bool aslr) {
-  return reserve_address_space_X(0, nth_bit(32), size, Metaspace::reserve_alignment(), aslr);
+  const size_t unscaled_max = nth_bit(narrow_klass_pointer_bits());
+  return reserve_address_space_X(0, unscaled_max, size, Metaspace::reserve_alignment(), aslr);
 }
 
 char* CompressedKlassPointers::reserve_address_space_for_zerobased_encoding(size_t size, bool aslr) {
-  return reserve_address_space_X(nth_bit(32), nth_bit(32 + LogKlassAlignmentInBytes), size, Metaspace::reserve_alignment(), aslr);
+  const size_t unscaled_max = nth_bit(narrow_klass_pointer_bits());
+  const size_t zerobased_max = nth_bit(narrow_klass_pointer_bits() + max_shift());
+  return reserve_address_space_X(unscaled_max, zerobased_max, size, Metaspace::reserve_alignment(), aslr);
 }
 
 char* CompressedKlassPointers::reserve_address_space_for_16bit_move(size_t size, bool aslr) {
   return reserve_address_space_X(nth_bit(32), nth_bit(48), size, nth_bit(32), aslr);
 }
 
-#if !defined(AARCH64) || defined(ZERO)
-// On aarch64 we have an own version; all other platforms use the default version
 void CompressedKlassPointers::initialize(address addr, size_t len) {
 
+  if (len > max_klass_range_size()) {
+    stringStream ss;
+    ss.print("Class space size (%zu) exceeds the maximum possible size (%zu)",
+              len, max_klass_range_size());
+    vm_exit_during_initialization(ss.base());
+  }
+
   // Remember the Klass range:
   _klass_range_start = addr;
   _klass_range_end = addr + len;
 
-  // The default version of this code tries, in order of preference:
-  // -unscaled    (base=0 shift=0)
-  // -zero-based  (base=0 shift>0)
-  // -nonzero-base (base>0 shift=0)
-  // Note that base>0 shift>0 should never be needed, since the klass range will
-  // never exceed 4GB.
-  constexpr uintptr_t unscaled_max = nth_bit(32);
-  assert(len <= unscaled_max, "Klass range larger than 32 bits?");
+  // Calculate Base and Shift:
 
-  constexpr uintptr_t zerobased_max = nth_bit(32 + LogKlassAlignmentInBytes);
+  if (UseCompactObjectHeaders) {
+
+    // In compact object header mode, with 22-bit narrowKlass, we don't attempt for
+    // zero-based mode. Instead, we set the base to the start of the klass range and
+    // then try for the smallest shift possible that still covers the whole range.
+    // The reason is that we want to avoid, if possible, shifts larger than
+    // a cacheline size.
+    _base = addr;
```

Comment:
Why is this important?

Reply:
It lessens the cache effects of Klass hyperaligning.

Reply:
Note that if we go with my KLUT proposal for post-Lilliput (the GC oop iteration improvements), this will not matter anymore and can be simplified to a fixed shift of 10.

Reply:
Yes, please, not having this code would be really nice.  This is difficult code.

Reply:
Do you seen any effects of this in anything other than special-crafted micro benchmarks? I wonder if it would be good enough to hard-code this to be 10 for the first integration of Lilliput.

Reply:
I will do some benchmarks

Reply:
I did SpecJBB runs with shift of 6, 8 and 10, respectively, which amounts to Klass alignment of 64, 256 and 1K. Benchmark scores did not show a significant pattern. I did not measure CPU stats though. 

But I still think a dynamically calculated shift makes sense, and I hesitate to change this code at this point. I therefore would like to move this question to followup RFEs if necessary.

Reply:
This code causes test errors in `CompressedClassPointersEncodingScheme.java` on s390 and PPC64. It forces the shift to `log_cacheline` which is 7 on PPC64 and 9 on s390. The test passes when we remove "s > log_cacheline && " from the condition below.
In addition, it doesn't fit to the comment which claims we should avoid shifts larger than the cacheline size. This enforces shifts to be larger than the cacheline size.

Reply:
> This code causes test errors in `CompressedClassPointersEncodingScheme.java` on s390 and PPC64. It forces the shift to `log_cacheline` which is 7 on PPC64 and 9 on s390. The test passes when we remove "s > log_cacheline && " from the condition below.

It's a bit late. We are close to pushing. While it should be harmless to drop below alignment to below cache line size, this would be a change affecting all platforms and would require all tests repeated. 

PPC/s390 are not targeted by the JEP. There had never been a discussion I am aware of that these platforms have to be clean with +COH. While it's nice that the changes had been contributed, I don't think that test errors on these platforms should hold up pushing this RFE. Therefore, if needed, we should just omit +COH part of the test for PPC/S390.

But then, what exactly is the error? If it's just the test assuming that cache line size is log 6, then the test should be fixed for ppc, not hotspot.

> In addition, it doesn't fit to the comment which claims we should avoid shifts larger than the cacheline size. This enforces shifts to be larger (or equal to) than the cacheline size.

?? The comment is correct. We try to avoid hyper alignment, hence we drop the shift to - if possible - log 2 cache line size. If it's equal to log 2 cache line size, we succeeded.

Reply:
PPC log2 will be `7` (`DATA_CACHE_SIZE = 128`) and for S390x it will be `8` (` = 256`). 

So I guess this change should be fine for now : 
```

Reply:
As I understand the comment, it says alignment <= cache line size. But the implementation makes alignment >= cache line size. "hyper alignment" means alignment > cache line size?
If we want alignment = cache line size, I'll be ok with disabling the +COH part of the test on PPC64 and s390. Correct, the problem is that the test assumes that log cache line size = 6.

Reply:
> As I understand the comment, it says alignment <= cache line size. But the implementation makes alignment >= cache line size. "hyper alignment" means alignment > cache line size?

Correct. since encoding range must cover the full klass range, and we only have 22bit nklass, shift is larger. at most 10. but since that causes hyper aligning, we try to get away with smaller shifts if klass range is smaller.